MaxText/inference/page_manager.py

```diff
@@ -1,232 +1,582 @@
+  """Represents the global state of memory pages managed by the `PageManager`.
+
+  This dataclass tracks the allocation status of each page across the entire system,
+  the mapping of pages to page groups (requests), and the current position within
+  each sequence's pages. State is managed globally, providing a single view
+  across all potential layers using this manager.
 
   Attributes:
-    page_status: Array indicating whether each page is in use (1) or free (0)
-    page_map: Array mapping slots to their assigned pages
-    sequence_lengths: Array containing the current length of each sequence
-    num_pages_used: Array tracking how many pages each slot is using
-    current_page: Array indicating the current active page for each slot
-    current_page_position: Array tracking position within current pages
+    page_status: A `jnp.ndarray` of shape `[num_pages]`. Each element
+      indicates whether the corresponding page in the global pool is free (0)
+      or allocated (1).
+    page_map: A `jnp.ndarray` of shape `[max_page_groups, max_pages_per_group]`.
+      This array maps each page group to the indices (within the global pool)
+      of its allocated pages. Entries beyond `num_pages_used` for a group are invalid.
+    num_pages_used: A `jnp.ndarray` of shape `[max_page_groups]`. This array
+      tracks the number of pages currently allocated to each page group. This
+      determines the valid entries in `page_map` for each group.
+    sequence_lengths: A `jnp.ndarray` of shape `[max_page_groups]`. This array
+      stores the current true length of each sequence (in tokens) associated
+      with a page group.
+    active_page: A `jnp.ndarray` of shape `[max_page_groups]`. This array
+      stores the global index of the *currently active* page (the page where the
+      next token will be written) for each page group. Only valid if the
+      corresponding `has_active_page` is True.
+    has_active_page: A `jnp.ndarray` of shape `[max_page_groups]`. Boolean mask
+      indicating whether a page group currently represents an active sequence
+      and thus whether its `active_page` and `active_page_position` entries
+      are meaningful.
+    active_page_position: A `jnp.ndarray` of shape `[max_page_groups]`. This array
+      stores the index (offset, 0 to tokens_per_page-1) of the next available
+      token *within the `active_page`* for each page group. Only valid if
+      `has_active_page` is True.
   """
 
-  page_status: Array
-  page_map: Array
-  sequence_lengths: Array
-  num_pages_used: Array
-  current_page: Array
-  current_page_position: Array
+  page_status: PagesInt1d
+  page_map: GroupsPagesInt2d
+  num_pages_used: GroupsInt1d
+  sequence_lengths: GroupsInt1d
+  active_page: GroupsInt1d
+  has_active_page: GroupsBool1d
+  active_page_position: GroupsInt1d
 
 
 def initialize_page_state(
     num_pages: int,
-    num_slots: int,
-    max_pages_per_slot: int,
+    max_page_groups: int,
```

Comment:
Curious can we use batch_size for max_page_groups?
concurrent number of sequences or requests is essentially the generate batch size.

Reply:
This is similar to one of Rupeng's earlier comments, which I addressed by removing the config variable `pagedattn_max_page_groups` from `base.yml`. So now inside `pyconfig.py` `max_page_groups = global_batch_size_to_load`. 

It could be made cleaner by removing the variable assignment in `pyconfig.py` and changing the line in `PageManager.__init__`

from 
`self.max_page_groups: int = config.pagedattn_max_page_groups`
to 
`self.max_page_groups: int = config.global_batch_size_to_load`.